Reprocessed Modal: Submission History

elipe17 opened the pull request:
## Summary of Changes
- Add ReprocessedModal component
- Use new component in submission history tables
Pull request closes #3271 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Submit some files.
3. Note that the `Reprocessed` button isn't visible in submission history
4. Reparse some/all of the files you submitted
5. You should now see the `Reprocessed` button is visible in submission history for the files you reparsed
6. Click the button and you should see a modal with the latest reparse date and explanation

In the `raft` environment use the following STT, year, and quarter to easily view the changes: `Arizona Year:2021 and Quarter:1`

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Submitted on/by columns are combined
- [x] Dialog has a close button
- [x] Dialog contains paragraph explainer on reprocessing
- [x] Dialog contains date time stamp for last reprocessing event
- [x] Dialog opens when trigger in Submitted on column is clicked
- [x] UX has been pinged to ticket knowledge center revisions to capture these changes
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [x] Does PR include documentation for Raft's a11y review? 
+ [x] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Add css class for reprocessed modal
- Add ReprocessedModal component
- Use new component for submission history tables

elipe17 committed:
- remove component

elipe17 committed:
- removed unnecessary span

elipe17 committed:
Merge branch 'develop' into 3271-reparse-dialog

elipe17 committed:
Small fix to rowspan

elipe17 committed:
- Add test for reprocessed button/modal

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 3271-reparse-dialog

elipe17 committed:
- linting

elipe17 committed:
- linting

victoriaatraft reviewed (approved):
<img width="1534" alt="Screenshot 2024-12-10 at 10 44 14 AM" src="https://github.com/user-attachments/assets/f4817b39-6630-45ef-8c39-c064105b0cb6">
LGTM from an a11y perspective. @elipe17. Is it possible for the modal to also close of the user selects the grayed out area in addition to the "close" button?

elipe17 committed:
- Add underline text decoration
- Fixed bug in Modal that caused it to crash when tabbing or shift tabbing with only one button

jtimpe commented on tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx:
```diff
@@ -26,13 +27,13 @@ const MonthSubRow = ({ data }) =>
         <th scope="rowgroup" rowSpan={3}>
           {formatDate(file.createdAt) + ' by ' + file.submittedBy}
-          {hasReparsed(file) && <></>}
+          {hasReparsed(file) && <ReprocessedModal date={reprocessedOn} />}
```
So this doesn't break anything because of how the modal's css/javascript work under the hood, but it is technically against DOM structure best practices to have an element appear visually "on top" but be nested inside other elements on the page. What you have would be how I would implement a tooltip component, but I'd recommend moving the modal to be a direct child of the fragment in `Reports.jsx` and letting the parent container own all the state, passing the `setState` function as a callback.

```jsx
export const Reports = () => {
  const [reprocessedModalVisible, setReprocessedModalVisible] = useState(false)
  // ......

  return (
    <>
      <div>
        ......
        <CaseAggregatesTable setReprocessedModalVisible={setReprocessedModalVisible} />
      </div>
      <Modal />
      <ReprocessedModal isVisible={reprocessedModalVisible} />
    </>
  )
}
```

`setReprocessedModalVisible` needs to be passed around a few times and may need to take some information like the reprocessed date (or the submission id, and the date can be fetched by the component). But would keep the page from creating a separate modal for each row. 

This line would then be changed to only contain the button, which executes the `setReprocessedModalVisible` callback

elipe17 replied:
Oh that is something I never considered. It did not occur to me that I was actually creating a modal per row in the tables. I will definitely make that change, thank you for the suggestion!

jtimpe replied:
@elipe17 @reitermb also while i'm thinking of it, tooltip might be available to us again after the v3 upgrade
https://designsystem.digital.gov/components/tooltip/

jtimpe commented on tdrs-frontend/src/components/SubmissionHistory/ReprocessedModal.jsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+const ReprocessedModal = ({ date }) => {
+  const [modalVisible, setModalVisible] = useState(false)
```
related to my other comment but wanted to call this out specifically - it's generally advised to keep state as high up in the component chain as makes sense. this goes along with having a single modal controlled by the root page. 
- https://legacy.reactjs.org/docs/lifting-state-up.html

jtimpe reviewed (commented):
Works as written! I have a suggestion regarding the structuring of the components that I would like your thoughts on before I approve, but I'll let you decide whether to move forward with it or not.

elipe17 committed:
- Moved model to parent component
- Add state props to update modal based on which files reprocessed button is clicked
- made the reprocessed button a separate component

elipe17 committed:
- Updated test

elipe17 committed:
- linting

elipe17 committed:
Merge branch 'develop' into 3271-reparse-dialog

elipe17 commented:
@jtimpe let me know what you think about the changes

elipe17 commented on tdrs-frontend/src/components/Modal/Modal.jsx:
```diff
@@ -24,18 +24,17 @@ const Modal = ({ title, message, buttons = [], isVisible = false }) => {
+      const btnIdxMinOne = Math.max(0, buttons.length - 1)
       if (shiftKey) {
         // go backward
+        const selectedIdxMinOne = Math.max(0, selectedButtonIndex - 1)
```
This fixed a bug where we could get an array out of bounds error. We could also escape the modal with tabs when this error occurred.

jtimpe reviewed (approved):
Good work!

raftmsohani reviewed (approved):
LGTM!

ADPennington commented on tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:
```diff
@@ -474,4 +486,42 @@ describe('SubmissionHistory', () => {
     }
   )
+
+  it('Shows Reprocessed button', () => {
```
@elipe17 do we need more tests here to capture the following scenarios:

- reprocessing tribal TANF file
- reprocessing SSP file
- no reprocessing 

Asking because I noticed that reprocessing indicators are not appearing for tribal TANF and SSP files. evidence below. 

![Screenshot 2024-12-16 124333](https://github.com/user-attachments/assets/bd30e3ff-1d81-4355-8dc2-f256c6b561ae)
![Screenshot 2024-12-16 123500](https://github.com/user-attachments/assets/4f7fa1d1-d84e-4b6d-9a66-c58ce8c13ad1)

elipe17 replied:
It should be the same exact logic regardless of TANF, SSP, Tribal. I just tested SSP locally and it was fine. Let me checkout the deployed env

elipe17 replied:
@ADPennington I just tested in the Raft env since it is still deployed there and it is working as expected for TANF, SSP, and Tribal. Maybe we can deploy this back to qasp when you are ready and check it out together?

ADPennington replied:
thank you @elipe17 -- one thing i forgot to mention is that I was testing aggregate and stratum files for SSP and tribal TANF.

elipe17 replied:
Roger. I tested active, closed, and aggregate for TANF. Active and aggregate for SPP. Only active for Tribal. Let me know when you are ready to test it out and we can hop on a call or work async via the deployment

ADPennington replied:
@elipe17 re-deploying now

ADPennington reviewed (changes requested):
changes requested re: reprocessing on tribal and SSP files.

elipe17 committed:
- do NOT consider an unfinished reparse event to be the latest reparse

ADPennington commented:
@ttran-hub -- do you have some time today for a11y testing for this PR? 

we need a11y testing on data file page >  submission history tab, in the areas highlighted in the screenshot below. This branch is in [qasp env](https://tdp-frontend-qasp.app.cloud.gov).

![Screenshot 2024-12-17 114759](https://github.com/user-attachments/assets/4eceb0ca-11ca-4c40-98f3-a7d53230996e)

@klinkoberstar can you remind me -- is there a release note associated with this?

ttran-hub commented:
> @ttran-hub -- do you have some time today for a11y testing for this PR?
> 
> we need a11y testing on data file page > submission history tab, in the areas highlighted in the screenshot below. This branch is in [qasp env](https://tdp-frontend-qasp.app.cloud.gov).
> 
> ![Screenshot 2024-12-17 114759](https://private-user-images.githubusercontent.com/63075587/396590603-4eceb0ca-11ca-4c40-98f3-a7d53230996e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0Njk0ODMsIm5iZiI6MTczNDQ2OTE4MywicGF0aCI6Ii82MzA3NTU4Ny8zOTY1OTA2MDMtNGVjZWIwY2EtMTFjYS00YzQwLTk4ZjMtYTdkNTMyMzA5OTZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDIwNTk0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3MjdkY2EzZWM1Y2IwYjdiYzMxZDgwNDcxNjc2MGFmNjlkMzdiZjg2YjI0MGM0ZGQyODA3MTA1OTBkOTFkMTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XleJCcuEs8Lz2tO4xkxI9yAHiBBczuUjjNd3GQkf9J4)
> 
> @klinkoberstar can you remind me -- is there a release note associated with this?

a111y LGTM!

ADPennington reviewed (approved):
🚀 🚀 🚀
